src/mofaflex/_core/pyro/likelihoods/base.py

```diff
@@ -0,0 +1,290 @@
+import inspect
+from abc import ABC, abstractmethod
+from collections.abc import Callable
+from inspect import isabstract
+
+import numpy as np
+import pyro
+import torch
+from numpy.typing import NDArray
+from pyro import distributions as dist
+from pyro.nn import PyroModule, PyroParam, PyroSample, pyro_method
+
+from ...utils import MeanStd
+
+EPS = 1e-8  # TODO: have some global definition/setting of EPS
+
+
+# https://stackoverflow.com/a/61350480
+class _PyroLikelihoodMeta(type(ABC), type(PyroModule)):
+    def __call__(cls, *args, **kwargs):
+        obj = cls.__new__(cls, *args, **kwargs)
+        args = list(args)
+        if obj.__class__ is not cls:
+            args = args[1:]
+        obj.__init__(*args, **kwargs)
+        return obj
+
+
+class PyroLikelihood(ABC, PyroModule, metaclass=_PyroLikelihoodMeta):
+    """Base class for MOFA-FLEX likelihoods used in the Pyro model.
+
+    Subclasses must implement `_model`, which returns a Pyro distribution object to be used as likelihood,
+    and `_guide`, which implements the variational distribution. Its return value is ignored.
+    """
+
+    __registry = {}
+
+    def __init__(
+        self,
+        view_name: str,
+        sample_dim: int,
+        feature_dim: int,
+        sample_means: dict[str, dict[str, NDArray[np.floating]]],
+        feature_means: dict[str, dict[str, NDArray[np.floating]]],
+    ):
+        """Initialize the Pyro likelihood.
+
+        Args:
+            view_name: The view (or guiding variable) name.
+            sample_dim: The sample dimension.
+            feature_dim: The feature dimension.
+            sample_means: Averages of samples across features for each group and view.
+            feature_means: Averages of features across samples for each group and view.
```

Comment:
Move this to the class documentation and remove the docstring for `__init__` to be consistent with the rest of the codebase.